src/common/protobuf_utils.hpp

```diff
@@ -64,6 +64,53 @@ struct Slave;
 
 namespace protobuf {
 
+//
+// A "protobuf union" is a protobuf message
+// which has an enum named "Type" defined in its scope,
+// a required field named "type" of a "Type" type
+// and a field corresponding to each member of the "Type" enum.
+//
+// A valid "protobuf union" should satisfy the following requirements.
+// 1. A member of the Type enum with a number (not index!) of 0,
+//    if there is any, should be named "UNKNOWN".
+// 2. For each other member of the Type enum there should be an optional field
+//    in the message with an exactly matching name in lowercase.
+// 3. The fields that correspond to members of the Type enum other than
+//    message.type(), should not be set.
+//
+// NOTE: If possible, oneof should be used in the new messages instead
+// of the "protobuf union".
+//
+class UnionValidator
+{
+public:
+  UnionValidator(
+      const google::protobuf::Descriptor*,
+      const google::protobuf::EnumDescriptor*);
+
+  Option<Error> validate(
+      const int messageTypeNumber, const google::protobuf::Message&) const;
+
+private:
+  std::vector<std::pair<int, const google::protobuf::FieldDescriptor*>>
+    unionFieldDescriptors_;
+  const google::protobuf::EnumDescriptor* typeDescriptor_;
+};
+
+
+// This function returns Error if the protobuf union is not valid
+// and Nothing otherwise.
+// In case MessageWithUnion is not a protobuf union,
+// this function will abort the process on the first use.
+template <class MessageWithUnion>
+Option<Error> validateProtobufUnion(const MessageWithUnion& message)
+{
+  static const UnionValidator validator(
+      MessageWithUnion::descriptor(), MessageWithUnion::Type_descriptor());
```

Comment:
Looking at this again, arriving at a `google::protobuf::EnumDescriptor` for the second argument can be easily missed, especially for those unfamiliar with protobuf generated code.  A note like this might help:
```
// NOTE: The `::Type_descriptor()` method is a protobuf generated method corresponding
// to the `Type` field of the given protobuf.  We expect this to be an enum field.
```

Reply:
I've just realized that the requirement that the field `type` has the type `Type` is 
a) not checked anywhere 
and 
b) is completely useless now, when the UnionValidator is initialized once per message type.
I'll simply remove this argument.

Reply:
I moved the enum descriptor extraction into the UnionValidator contructor. Now it is much more local, and I hope no comment is needed to understand what is going on there.